### ISSUE TITLE
SDK v2 - After a _loadFromJson call, under the _postCreate method a wrong UID is assigned

### ISSUE THREAD

**Issue by sotano42** [Scirra/Construct-bugs#8261]
## Problem description

You have N instances of an object using a beavior. You save the game state with those instances, then destroy one. Lastly, you load the game state to recreate the instances, you will notice at the _postCreate metod the UID of the new instance is incorrect. It contains the last UID available, not the newest that should be assigned.

May be related to issue #8209

## Attach a .c3p

Addon: [behavior-basic-bullet-postCreate UID Error.zip](https://github.com/user-attachments/files/17201281/behavior-basic-bullet-postCreate.UID.Error.zip)
Project: [test-save-wrongUID.zip](https://github.com/user-attachments/files/17201296/test-save-wrongUID.zip)

## Steps to reproduce

Install the attached addon (basic bullet sample), and run the attached project as follows:

1. Run the attached project, open the dev console and notice the "Post create called:" messages with the created UIDs
![image](https://github.com/user-attachments/assets/20d313f5-2d46-40a3-a7af-d9c04b9c6403)

2. Press "S" to save the game state
3. Press "D" to destroy UID 2.
4. Press "L" to load the state
5. Notice when the new instance is created to rebuild the destroyed instance with UID 2 it logs the message "Post create called:" referencing UID 4 which is incorrect, because that UID is already assigned to an exising instance.
![image](https://github.com/user-attachments/assets/0a99759c-014d-450f-bff0-c2f1de058d48)

## Observed result

An incorrect UID is referenced for the newly created instance.

## Expected result

A new UID should be assigned, not an existing one.

## More details

Stable r407.2
Beta r408

## System details

No exception thrown, but incorrect UID is assigned.

**Comment by AshleyScirra** [comment]
I think it is in fact working correctly. When loading a savegame the engine may either load state for existing instances, or create a new instance. So following these steps, the engine may do the following:

1. Upon saving, saves state for instances UID 2, 3 and 4
2. Destroy instance UID 2 so only 3 and 4 are left
3. Upon loading, restore state as follows:
    - existing instance UID 3: loaded as instance UID 2
    - existing instance UID 4: loaded as instance UID 3
    - new instance created with UID 4 - which then triggers _postCreate()

In other words the UIDs may change when loading a savegame, and the engine will load data from the savegame in to existing instances first, before it proceeds to creating new instances.

**Comment by sotano42** [comment]
> * which then triggers _postCreate()

I'm pretty sure this is wrong, because the postCreate is only triggered for the new instance created. If we access the UID from the _postCreate, it should return the correct UID, which in this case is not matching. 

Run the project again in debug mode and you will see that the UID created doesn't match the UID available at _postCreate.

### Before destroy:in UID 2
![image](https://github.com/user-attachments/assets/03193c69-a6e0-4057-b086-f7ae198f3520)
![image](https://github.com/user-attachments/assets/0b45089c-1e47-4aa5-b366-b08ccf0e2e74)

### After the destroy UID 2, only 3 and 4 are left:
![image](https://github.com/user-attachments/assets/b138cc2a-dc29-4b7e-bdd8-272fc1ff3e10)
![image](https://github.com/user-attachments/assets/71f3387d-6b05-4e5c-b465-bc4c177320ed)

### Upon Load, the supposed assigned UID was 4 at post create, but the UID created is actually UID 2:
![image](https://github.com/user-attachments/assets/2aac1db3-b840-45b9-af58-de7ec4de0ebc)
![image](https://github.com/user-attachments/assets/d008e93f-cc1f-4e33-b2a7-b6fffcf005e0)
![image](https://github.com/user-attachments/assets/f97749e5-e593-4df0-9240-77c1f7aec0dd)

Post create was only called for the instance created
![image](https://github.com/user-attachments/assets/f7ac7d4e-c2f6-471b-bd9d-57cccfe7a0be)

@AshleyScirra **If the UID changes, that's totally fine, but the postCreate should return the correct one.** My addons rely on the UID, it should match the correct instance.

**Comment by AshleyScirra** [comment]
I don't see how this is any different to my previous comment. It is possible, and valid, for the runtime to create the instance with UID 4 in this scenario.

**Comment by sotano42** [comment]
@AshleyScirra I'm not discussing the UID change but instead that the UID doesn't match the real that was created. UiD 2 was created to rebuild the Instance, but postCreate returns UID 4 which doesn't match.

**Comment by sotano42** [comment]
@AshleyScirra  here's some constructive thought on why I insist this is wrong.

Whenever an instance is loaded Construct reutilizes what's there to recreate it, which theoretically is "disposing" the instance and using it potentially to represent a different UID. This is inconsistent, because the original UID 2, can potentially be any other number, but the _release methods were never called, thus I can't let my framework know that the instance is not the same. 
On that scenenario I will have to consider that any instance loaded is already invalid and I should dispose it, which will cause me to re-initialize whatever I hade previously attached to it. 

If an instance is reloaded, it shouldn't be assigned to a new UID, otherwise what's the purpose of reutilizing it if the reuse process will cause to consider it invalid and re-attach the functionality to the plugin?

From my perspective, it should work like this to be consistent:

### Alternative 1: The instance is reassigned to a different UID and a dispose/release method is called.
1. If Construct will still use the same instance to assign a new UID, then the _release method should be called (or similar) to let the dev know the instance is no longer valid.
2. The custom functionality will dispose whatever it needs because the _release method already marked the instance as invalid, avoiding further inconsistencies.
3. After the release is called, you complete the instance initialization and then call the _postCreate to trigger the custom functionality.
4. Custom functionality will be built again considering the new UID. This will cause an overhead, which is what I believe you tried to avoid by reusing the same instance, however now the overhead is passed to the custom functionality.

### Alternative 2: The instance is not reassigned to a different UID.
1. Construct must ensure the same UID is reutilized to represent the object.
2. _postCreate method will be called to complete teh custom functionality.
3. The custom functionality will update the instance considering it represents the same object, reducing the overhead of recreating it again and being consistent with the performance optimization of reusing the same UID.

If none of the above are implemented, it will mean that every time the _loadFromJson method is called we should consider the instance as disposed/invalid/destroyed and retrigger all the custom initializations again.

**Comment by sotano42** [comment]
### Current State: Instance is never marked as invalid/disposed/destroyed, but the UID is reassigned.

@AshleyScirra This is the kind of workaround we'll have to implement due to the UID reassignment, which I find inconsistent and may not solve all the cases. 

My instances had a reference stored in _button, pointing to the instance that was just loaded without being destroyed. However, the UID is now different and represents a completely different object. The result is that the loaded instance is always invalid and will never match the reference I had previously, meaning it must be recreated from scratch. This contradicts the idea of having a unique UID, which is meant to identify a single object for its lifetime. Furthermore, other custom properties will always be inconsistent unless manually reset, I'm not sure other devs know about this restriction.

UID should always match if the instance is never destroyed. 

![image](https://github.com/user-attachments/assets/c0b14163-6eda-438c-b83e-ac42ea5d7c19)

**Comment by AshleyScirra** [comment]
I think basically this is all working by design in terms of the Construct engine, but the way it works is a bit awkward for your particular addon. I don't think that qualifies as a bug. You could post a feature request to change how it works, but this is a complex and old part of Construct (it's worked like this for years) and changing it could have other unexpected backwards compatibility problems - such as breaking other addons that depend on UIDs working as they do now. I think the most pragmatic approach is to work around it, either as you appear to have done so already, or with the aid of new SDK features - for example there's already an "afterload" event, but perhaps adding a "beforeload" event would provide a point at which the existing UID state could be remembered and then re-applied in "afterload" to work around changing UIDs.
